README
ELF prelinking utility to speed up dynamic linking.


ChangeLog
2013-10-05  Jakub Jelinek  <jakub@redhat.com>

	* src/arch-s390.c (s390_prelink_conflict_rela): For R_390_IRELATIVE,
	always use that relocation type in the conflict section.

2013-10-01  Jakub Jelinek  <jakub@redhat.com>

	* testsuite/ifunc.h (IFUNC_ASM): Add s390x and s390 version.

2013-09-30  Richard Sandiford  <rsandifo@linux.vnet.ibm.com>

	* src/arch-s390.c (s390_prelink_conflict_rela): Fix handling of
	R_390_IRELATIVE.
	* src/arch-s390x.c (s390x_prelink_conflict_rela): Likewise.

2013-07-19  Marek Polacek  <polacek@redhat.com>

	* doc/prelink.8 (--usage): Document it. 
	(--layout-page-size=): Likewise.

2013-05-03  Jakub Jelinek  <jakub@redhat.com>

	* src/cache.c (prelink_save_cache_check): For ET_CACHE_DYN
	or ET_UNPRELINKABLE dependencies, call prelink_save_cache_check
	recursively.
	* testsuite/Makefile.am (TESTS): Add unprel1.sh.
	* testsuite/Makefile.in: Regenerated.
	* testsuite/unprel1.sh: New test.
	* testsuite/unprel1.c: New file.
	* testsuite/unprel1lib1.c: New file.
	* testsuite/unprel1lib2.c: New file.

	* src/prelink.h (R_390_IRELATIVE): Define if not defined.

2013-04-25  Jakub Jelinek  <jakub@redhat.com>

	* src/arch-s390.c (s390_adjust_rela, s390_prelink_rela,
	s390_apply_conflict_rela, s390_prelink_conflict_rela,
	s390_undo_prelink_rela): Handle R_390_IRELATIVE.
	* src/arch-s390x.c (s390x_adjust_rela, s390x_prelink_rela,
	s390x_apply_conflict_rela, s390x_prelink_conflict_rela,
	s390x_undo_prelink_rela, s390x_reloc_size): Likewise.

2013-04-18  Jakub Jelinek  <jakub@redhat.com>

	* testsuite/reloc4.sh: Add libraries to link lines to make
	testsuite work both with --no-add-needed and with --add-needed,
	including latest ld.bfd.
	* testsuite/reloc5.sh: Likewise.
	* testsuite/reloc8.sh: Likewise.
	* testsuite/reloc9.sh: Likewise.

2013-04-17  Jakub Jelinek  <jakub@redhat.com>

	* testsuite/ifunc.h (IFUNC_DECL): Fix up .size on ppc64.
	* testsuite/cxx1.sh: Add libraries to link lines to make
	testsuite work both with --no-add-needed and with --add-needed,
	including latest ld.bfd.
	* testsuite/cxx2.sh: Likewise.
	* testsuite/cxx3.sh: Likewise.
	* testsuite/deps1.sh: Likewise.
	* testsuite/deps2.sh: Likewise.
	* testsuite/ifunc1.sh: Likewise.
	* testsuite/ifunc2.sh: Likewise.
	* testsuite/ifunc3.sh: Likewise.
	* testsuite/quick1.sh: Likewise.
	* testsuite/quick2.sh: Likewise.
	* testsuite/quick3.sh: Likewise.
	* testsuite/reloc1.sh: Likewise.
	* testsuite/reloc10.sh: Likewise.
	* testsuite/reloc11.sh: Likewise.
	* testsuite/reloc3.sh: Likewise.
	* testsuite/reloc6.sh: Likewise.
	* testsuite/reloc7.sh: Likewise.
	* testsuite/shuffle1.sh: Likewise.
	* testsuite/shuffle2.sh: Likewise.
	* testsuite/shuffle3.sh: Likewise.
	* testsuite/shuffle4.sh: Likewise.
	* testsuite/shuffle5.sh: Likewise.
	* testsuite/shuffle6.sh: Likewise.
	* testsuite/shuffle7.sh: Likewise.
	* testsuite/shuffle8.sh: Likewise.
	* testsuite/shuffle9.sh: Likewise.
	* testsuite/tls1.sh: Likewise.
	* testsuite/tls2.sh: Likewise.
	* testsuite/tls3.sh: Likewise.
	* testsuite/tls4.sh: Likewise.
	* testsuite/tls5.sh: Likewise.
	* testsuite/tls6.sh: Likewise.
	* testsuite/undo1.sh: Likewise.

2012-06-28  Jakub Jelinek  <jakub@redhat.com>

	* src/get.c (is_ldso_soname): Add arm dynamic linker SONAMEs.

	* src/dwarf2.h (DW_FORM_GNU_ref_alt,
	DW_FORM_GNU_strp_alt): Define.
	* src/dwarf2.c (read_abbrev): Allow those forms.
	(adjust_attributes): Handle them.

2012-04-10  Jakub Jelinek  <jakub@redhat.com>

	* src/dwarf2.c (DEBUG_MACRO): Define.
	(debug_sections): Add .debug_macro entry.

2012-04-02  Jakub Jelinek  <jakub@redhat.com>

	* src/dwarf2.c (adjust_attributes): For DWARF4+ treat
	DW_AT_data_member_location with DW_FORM_data[48] as
	large constant instead of loclistptr.

2011-10-12  Jakub Jelinek  <jakub@redhat.com>

	* src/main.c (layout_page_size): New variable.
	(OPT_LAYOUT_PAGE_SIZE): Define.
	(options): Add --layout-page-size option.
	(set_default_layout_page_size): New function in i?86/x86_64
	resp. empty macro on other architectures.
	(main): Call it.
	* src/arch-i386.c (i386_layout_libs_init): Fail if
	l->max_page_size is larger than 2MB and exec_shield is true.
	Compute reg0s as REG0S rounded up to l->max_page_size
	boundary and use it instead of REG0S.
	(i386_layout_libs_pre): Compute reg0s as REG0S rounded up to
	l->max_page_size boundary and use it instead of REG0S.
	Fix up a pasto in computation of virt parameter for
	third i386_find_free_addr call.
	* src/prelink.h (layout_page_size): New decl.
	* src/layout.c (layout_libs): Set max_page_size and
	l.max_page_size to maximum of plarch->max_page_size
	and layout_page_size.  Fail if mmap_base or mmap_end
	aren't max_page_size aligned.
	* src/arch-ppc.c (ppc_layout_libs_pre): Give up if
	l->max_page_size is larger than 64KB.

2011-08-26  Jakub Jelinek  <jakub@redhat.com>

	* testsuite/cxx3.sh: Expect fewer conflicts.

	* src/arch-arm.c (PL_ARCH): Change .dynamic_linker to
	/lib/ld-linux.so.3.

2011-08-25  Jakub Jelinek  <jakub@redhat.com>

	* src/gather.c (gather_func): When looking for PT_PHDR header,
	skip any PT_LOPROC ... PR_HIPROC phdrs.

	* testsuite/cxx1.sh: Don't check for any "Removing C++ conflict"
	messages on arm.
	* testsuite/cxx2.sh: Likewise.
	* testsuite/cxx3.sh: New test.
	* testsuite/cxx3.C: New file.
	* testsuite/cxx3.h: New file.
	* testsuite/cxx3lib1.C: New file.
	* testsuite/cxx3lib2.C: New file.
	* testsuite/Makefile.am (TESTS): Add cxx3.sh.
	* testsuite/Makefile.in: Regenerated.

2011-08-25  William Cohen  <wcohen@redhat.com>

	* testsuite/shuffle2.c: Use %nobits instead of @nobits
	to be arm assembler friendly.
	* testsuite/shuffle9.c: Likewise.
	* testsuite/reloc5.c (main): Likewise.

2011-06-22  Jakub Jelinek  <jakub@redhat.com>

	* src/dwarf2.h (DW_OP_GNU_parameter_ref): Define.
	* src/dwarf2.c (adjust_location_list): Handle
	DW_OP_GNU_parameter_ref.

2011-05-11  Jakub Jelinek  <jakub@redhat.com>

	* src/dwarf2.h (DW_OP_GNU_const_type, DW_OP_GNU_regval_type,

src/sha.c
/* sha.c - Functions to compute the SHA1 hash (message-digest) of files
   or blocks of memory.  Complies to the NIST specification FIPS-180-1.

   Copyright (C) 2000, 2001, 2003 Scott G. Miller

   Credits:
      Robert Klep <robert@ilse.nl>  -- Expansion function fix
   NOTE: The canonical source of this file is maintained in GNU coreutils.
*/

#include <sys/types.h>
#include <stdlib.h>
#include <string.h>
#include <byteswap.h>
#include "md5.h"
#include "sha.h"

/*
  Not-swap is a macro that does an endian swap on architectures that are
  big-endian, as SHA needs some data in a little-endian format
*/

#if __BYTE_ORDER == __BIG_ENDIAN
# define SWAP(n)  bswap_32 (n)
# define NOTSWAP(n) (n)
#else
# define SWAP(n) (n)
# define NOTSWAP(n) bswap_32 (n)
#endif

/* This array contains the bytes used to pad the buffer to the next
   64-byte boundary.  (RFC 1321, 3.1: Step 1)  */
static const unsigned char fillbuf[64] = { 0x80, 0 /* , 0, 0, ...  */ };


/*
  Takes a pointer to a 160 bit block of data (five 32 bit ints) and
  intializes it to the start constants of the SHA1 algorithm.  This
  must be called before using hash in the call to sha_hash
*/
void
sha_init_ctx (struct sha_ctx *ctx)
{
  ctx->A = 0x67452301;
  ctx->B = 0xefcdab89;
  ctx->C = 0x98badcfe;
  ctx->D = 0x10325476;
  ctx->E = 0xc3d2e1f0;

  ctx->total[0] = ctx->total[1] = 0;
  ctx->buflen = 0;
}

/* Put result from CTX in first 20 bytes following RESBUF.  The result
   must be in little endian byte order.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
sha_read_ctx (const struct sha_ctx *ctx, void *resbuf)
{
  ((md5_uint32 *) resbuf)[0] = NOTSWAP (ctx->A);
  ((md5_uint32 *) resbuf)[1] = NOTSWAP (ctx->B);
  ((md5_uint32 *) resbuf)[2] = NOTSWAP (ctx->C);
  ((md5_uint32 *) resbuf)[3] = NOTSWAP (ctx->D);
  ((md5_uint32 *) resbuf)[4] = NOTSWAP (ctx->E);

  return resbuf;
}

/* Process the remaining bytes in the internal buffer and the usual
   prolog according to the standard and write the result to RESBUF.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
sha_finish_ctx (struct sha_ctx *ctx, void *resbuf)
{
  /* Take yet unprocessed bytes into account.  */
  md5_uint32 bytes = ctx->buflen;
  size_t pad;

  /* Now count remaining bytes.  */
  ctx->total[0] += bytes;
  if (ctx->total[0] < bytes)
    ++ctx->total[1];

  pad = bytes >= 56 ? 64 + 56 - bytes : 56 - bytes;
  memcpy (&ctx->buffer[bytes], fillbuf, pad);

  /* Put the 64-bit file length in *bits* at the end of the buffer.  */
  *(md5_uint32 *) &ctx->buffer[bytes + pad + 4] = NOTSWAP (ctx->total[0] << 3);
  *(md5_uint32 *) &ctx->buffer[bytes + pad] = NOTSWAP ((ctx->total[1] << 3) |
						    (ctx->total[0] >> 29));

  /* Process last bytes.  */
  sha_process_block (ctx->buffer, bytes + pad + 8, ctx);

  return sha_read_ctx (ctx, resbuf);
}

/* Compute MD5 message digest for LEN bytes beginning at BUFFER.  The
   result is always in little endian byte order, so that a byte-wise
   output yields to the wanted ASCII representation of the message
   digest.  */
void *
sha_buffer (const char *buffer, size_t len, void *resblock)
{
  struct sha_ctx ctx;

  /* Initialize the computation context.  */
  sha_init_ctx (&ctx);

  /* Process whole buffer but last len % 64 bytes.  */
  sha_process_bytes (buffer, len, &ctx);

  /* Put result in desired memory area.  */
  return sha_finish_ctx (&ctx, resblock);
}

void
sha_process_bytes (const void *buffer, size_t len, struct sha_ctx *ctx)
{
  /* When we already have some bits in our internal buffer concatenate
     both inputs first.  */
  if (ctx->buflen != 0)
    {
      size_t left_over = ctx->buflen;
      size_t add = 128 - left_over > len ? len : 128 - left_over;

      memcpy (&ctx->buffer[left_over], buffer, add);
      ctx->buflen += add;

      if (ctx->buflen > 64)
	{
	  sha_process_block (ctx->buffer, ctx->buflen & ~63, ctx);

	  ctx->buflen &= 63;
	  /* The regions in the following copy operation cannot overlap.  */
	  memcpy (ctx->buffer, &ctx->buffer[(left_over + add) & ~63],
		  ctx->buflen);
	}

      buffer = (const char *) buffer + add;
      len -= add;
    }

  /* Process available complete blocks.  */
  if (len >= 64)
    {
#define UNALIGNED_P(p) (((md5_uintptr) p) % __alignof__ (md5_uint32) != 0)
      if (UNALIGNED_P (buffer))
	while (len > 64)
	  {
	    sha_process_block (memcpy (ctx->buffer, buffer, 64), 64, ctx);
	    buffer = (const char *) buffer + 64;
	    len -= 64;
	  }
      else
	{
	  sha_process_block (buffer, len & ~63, ctx);
	  buffer = (const char *) buffer + (len & ~63);
	  len &= 63;
	}
    }

  /* Move remaining bytes in internal buffer.  */
  if (len > 0)
    {
      size_t left_over = ctx->buflen;

      memcpy (&ctx->buffer[left_over], buffer, len);
      left_over += len;
      if (left_over >= 64)
	{
	  sha_process_block (ctx->buffer, 64, ctx);
	  left_over -= 64;
	  memcpy (ctx->buffer, &ctx->buffer[64], left_over);
	}
      ctx->buflen = left_over;
    }
}

/* --- Code below is the primary difference between md5.c and sha.c --- */

/* SHA1 round constants */
#define K1 0x5a827999L
#define K2 0x6ed9eba1L
#define K3 0x8f1bbcdcL
#define K4 0xca62c1d6L

/* Round functions.  Note that F2 is the same as F4.  */
#define F1(B,C,D) ( D ^ ( B & ( C ^ D ) ) )
#define F2(B,C,D) (B ^ C ^ D)
#define F3(B,C,D) ( ( B & C ) | ( D & ( B | C ) ) )
#define F4(B,C,D) (B ^ C ^ D)

/* Process LEN bytes of BUFFER, accumulating context into CTX.
   It is assumed that LEN % 64 == 0.
   Most of this code comes from GnuPG's cipher/sha1.c.  */

void
sha_process_block (const void *buffer, size_t len, struct sha_ctx *ctx)
{
  const md5_uint32 *words = buffer;
  size_t nwords = len / sizeof (md5_uint32);
  const md5_uint32 *endp = words + nwords;
  md5_uint32 x[16];
  md5_uint32 a = ctx->A;
  md5_uint32 b = ctx->B;
  md5_uint32 c = ctx->C;
  md5_uint32 d = ctx->D;
  md5_uint32 e = ctx->E;

  /* First increment the byte count.  RFC 1321 specifies the possible
     length of the file up to 2^64 bits.  Here we only compute the
     number of bytes.  Do a double word increment.  */
  ctx->total[0] += len;
  if (ctx->total[0] < len)
    ++ctx->total[1];

#define M(I) ( tm =   x[I&0x0f] ^ x[(I-14)&0x0f] \
		    ^ x[(I-8)&0x0f] ^ x[(I-3)&0x0f] \
	       , (x[I&0x0f] = rol(tm, 1)) )

#define R(A,B,C,D,E,F,K,M)  do { E += rol( A, 5 )     \
				      + F( B, C, D )  \
				      + K	      \
				      + M;	      \
				 B = rol( B, 30 );    \
			       } while(0)

  while (words < endp)
    {
      md5_uint32 tm;
      int t;
      /* FIXME: see sha1.c for a better implementation.  */
      for (t = 0; t < 16; t++)
	{
	  x[t] = NOTSWAP (*words);
	  words++;
	}

      R( a, b, c, d, e, F1, K1, x[ 0] );
      R( e, a, b, c, d, F1, K1, x[ 1] );
      R( d, e, a, b, c, F1, K1, x[ 2] );
      R( c, d, e, a, b, F1, K1, x[ 3] );
      R( b, c, d, e, a, F1, K1, x[ 4] );
      R( a, b, c, d, e, F1, K1, x[ 5] );
      R( e, a, b, c, d, F1, K1, x[ 6] );
      R( d, e, a, b, c, F1, K1, x[ 7] );
      R( c, d, e, a, b, F1, K1, x[ 8] );
      R( b, c, d, e, a, F1, K1, x[ 9] );
      R( a, b, c, d, e, F1, K1, x[10] );
      R( e, a, b, c, d, F1, K1, x[11] );
      R( d, e, a, b, c, F1, K1, x[12] );
      R( c, d, e, a, b, F1, K1, x[13] );
      R( b, c, d, e, a, F1, K1, x[14] );
      R( a, b, c, d, e, F1, K1, x[15] );
      R( e, a, b, c, d, F1, K1, M(16) );
      R( d, e, a, b, c, F1, K1, M(17) );
      R( c, d, e, a, b, F1, K1, M(18) );
      R( b, c, d, e, a, F1, K1, M(19) );
      R( a, b, c, d, e, F2, K2, M(20) );
      R( e, a, b, c, d, F2, K2, M(21) );
      R( d, e, a, b, c, F2, K2, M(22) );
      R( c, d, e, a, b, F2, K2, M(23) );
      R( b, c, d, e, a, F2, K2, M(24) );
      R( a, b, c, d, e, F2, K2, M(25) );
      R( e, a, b, c, d, F2, K2, M(26) );
      R( d, e, a, b, c, F2, K2, M(27) );
      R( c, d, e, a, b, F2, K2, M(28) );
      R( b, c, d, e, a, F2, K2, M(29) );
      R( a, b, c, d, e, F2, K2, M(30) );
      R( e, a, b, c, d, F2, K2, M(31) );
      R( d, e, a, b, c, F2, K2, M(32) );
      R( c, d, e, a, b, F2, K2, M(33) );
      R( b, c, d, e, a, F2, K2, M(34) );
      R( a, b, c, d, e, F2, K2, M(35) );
      R( e, a, b, c, d, F2, K2, M(36) );
      R( d, e, a, b, c, F2, K2, M(37) );
      R( c, d, e, a, b, F2, K2, M(38) );
      R( b, c, d, e, a, F2, K2, M(39) );
      R( a, b, c, d, e, F3, K3, M(40) );
      R( e, a, b, c, d, F3, K3, M(41) );
      R( d, e, a, b, c, F3, K3, M(42) );
      R( c, d, e, a, b, F3, K3, M(43) );
      R( b, c, d, e, a, F3, K3, M(44) );
      R( a, b, c, d, e, F3, K3, M(45) );
      R( e, a, b, c, d, F3, K3, M(46) );
      R( d, e, a, b, c, F3, K3, M(47) );
      R( c, d, e, a, b, F3, K3, M(48) );
      R( b, c, d, e, a, F3, K3, M(49) );
      R( a, b, c, d, e, F3, K3, M(50) );
      R( e, a, b, c, d, F3, K3, M(51) );
      R( d, e, a, b, c, F3, K3, M(52) );
      R( c, d, e, a, b, F3, K3, M(53) );
      R( b, c, d, e, a, F3, K3, M(54) );
      R( a, b, c, d, e, F3, K3, M(55) );
      R( e, a, b, c, d, F3, K3, M(56) );
      R( d, e, a, b, c, F3, K3, M(57) );
      R( c, d, e, a, b, F3, K3, M(58) );
      R( b, c, d, e, a, F3, K3, M(59) );
      R( a, b, c, d, e, F4, K4, M(60) );
      R( e, a, b, c, d, F4, K4, M(61) );
      R( d, e, a, b, c, F4, K4, M(62) );
      R( c, d, e, a, b, F4, K4, M(63) );
      R( b, c, d, e, a, F4, K4, M(64) );
      R( a, b, c, d, e, F4, K4, M(65) );
      R( e, a, b, c, d, F4, K4, M(66) );
      R( d, e, a, b, c, F4, K4, M(67) );
      R( c, d, e, a, b, F4, K4, M(68) );
      R( b, c, d, e, a, F4, K4, M(69) );
      R( a, b, c, d, e, F4, K4, M(70) );
      R( e, a, b, c, d, F4, K4, M(71) );
      R( d, e, a, b, c, F4, K4, M(72) );
      R( c, d, e, a, b, F4, K4, M(73) );
      R( b, c, d, e, a, F4, K4, M(74) );
      R( a, b, c, d, e, F4, K4, M(75) );
      R( e, a, b, c, d, F4, K4, M(76) );
      R( d, e, a, b, c, F4, K4, M(77) );
      R( c, d, e, a, b, F4, K4, M(78) );
      R( b, c, d, e, a, F4, K4, M(79) );

      a = ctx->A += a;
      b = ctx->B += b;
      c = ctx->C += c;
      d = ctx->D += d;
      e = ctx->E += e;
    }
}
